### PR TITLE
[DS-3498] Full Text Index Behavior on Public Items with Embargoed Bitstreams

### PR DESCRIPTION
**Pull request by jsavell**
https://jira.duraspace.org/browse/DS-3498

 - Verify pdf bitstream is publicly accessible before indexing

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/discovery/FullTextContentStreams.java [hunk 1]
```diff
@@ -65,27 +71,41 @@ protected void init(Item parentItem) {
...
                             + fulltextBitstream.getName());
                 }
             }
         }
     }
+    
+    private boolean isIndexable(Bitstream fulltextBitstream,Bundle myBundle) throws SQLException {
+        for (ResourcePolicy rp:bundleService.getBitstreamPolicies(context, myBundle)) {
+            if (rp.getdSpaceObject().getID() == fulltextBitstream.getID() && rp.getGroup().getName().toLowerCase().equals("anonymous") 
+                    && resourcePolicyService.isDateValid(rp)) {
```
Comment (tomdesair) [inline dspace-api/src/main/java/org/dspace/discovery/FullTextContentStreams.java]: Evaluating the resource policy at the moment of indexing might cause problems (see general comment).

**Review by tomdesair — CHANGES_REQUESTED** [review]
Thanks for the pull-request @jsavell. This is not an easy problem to fix and I think there are some issues with your approach:

* What happens if the embargo on a bitstream expires? It will not be index automatically because currently there is no trigger when an embargo expires (-> the resource policies becomes valid). Your current solution would require the repository to re-evaluate each bitstream every night.

* Not indexing the text at all will also prevent Administrator or users with the correct group membership from searching on the full text of an embargoed bitstream. With this implementation the full text search feature will be completely broken for private repositories (each bitstream is restricted to a group != anonymous).

File: dspace-api/src/main/java/org/dspace/discovery/FullTextContentStreams.java (hunk 1 above)
Comment (jsavell) [inline dspace-api/src/main/java/org/dspace/discovery/FullTextContentStreams.java]: Okay. Our use case at A&M was more straightforward. We simply didn't want the texts indexed if they weren't also publicly available. And we do update our SOLR index nightly.

The short term advice in the JIRA issue for those concerned was to just turn snippet indexing off altogether while a solution was researched. Would adding an on/off configuration parameter to this implementation make it more broadly useful?

For the second bullet point, could this be solved by allowing the policy names to be excluded/allowed to also be set through configuration?

**Comment by tdonohue** [comment]
Revisiting this PR today (as I see this core ticket has not yet been resolved). As-is, it looks like this PR attempts to solve the issue by only indexing Bitstreams which are publicly accessible (i.e. anonymous access is currently available). 

While this "works" for some use cases, unfortunately, this has a negative side effect on Administrative search functionality within DSpace. Essentially, with this PR in place, DSpace's ["Access Rights Awareness" feature](https://wiki.duraspace.org/display/DSDOC6x/Discovery#Discovery-AccessRightsAwareness)  will no longer work for full text searching, as access restricted content won't be indexed (at all). So, any DSpace sites which have large amounts of access-restricted content will be unable to search within that content when logged in as an Administrator.

As noted by @tomdesair above in his comment, an additional difficulty with this implementation is that it would require regularly reindexing all content within your site (e.g. on a nightly or weekly basis). Currently, in this implementation, when an embargo expires, the content will remain hidden in search results (until the next full reindex of content takes place).

So, this fix could be applied as a "quick fix" for some scenarios / use cases (assuming neither of the above issues is of concern to your local DSpace needs), but it's not an ideal long-term fix.

Looking back at the ticket [DS-3498](https://jira.duraspace.org/browse/DS-3498), we might want to think about a broader "quick fix" simply being to disable all hit highlighting / search snippets *by default*.  It seems the larger issue here is that bitstreams/files don't have their access rights stored in Discovery/Solr, so the search results are not able to take them into account.

**Comment by tdonohue** [comment]
Alternative quick fix which simply disables full-text search snippets by default: #2069  (Obviously, this alternative PR is also not a long term fix. But, it resolves the problem without affecting ability to search access restricted content.)

**Pull request closed without merging**